experimental/segmenter/src/math_helper.rs

```diff
@@ -2,59 +2,288 @@
+        // for i in 0..b.dim().0 {
+        //     self.submatrix_mut::<1>(i).add_dot_2d(a, b.submatrix(i));
+        // }
+        let lhs = a.as_slice();
+        for i in 0..n {
+            if let (Some(dest), Some(rhs)) = (
+                self.as_mut_slice().get_mut(i),
+                b.as_slice().get(i * m..(i + 1) * m),
+            ) {
+                *dest += unrolled_dot(lhs, rhs);
+            } else {
+                debug_assert!(false);
+            }
+        }
+    }
 }
 
-/// `tanh_arr1` computes elementwise tanh function for elements of a 1d array
-pub fn tanh_arr1(arr: ArrayBase<ViewRepr<&f32>, Dim<[usize; 1]>>) -> Array1<f32> {
-    arr.map(|v| v.tanh())
+// Polyfill float operations with libm in case we're no_std.
+#[allow(unused_imports)]
+use num_traits::Float;
+
+/// `tanh` computes the tanh function for a scalar value.
+#[inline]
+pub fn tanh(x: f32) -> f32 {
+    x.tanh()
 }
 
-/// `change_row` gets one 2d array (`arr`), one 1d array (`arr1`), and an index (`row_id`), and replaces the `row_id`-th row of
-/// `arr` with `arr1`
-pub fn change_row(mut arr: Array2<f32>, row_id: usize, new_row: &Array1<f32>) -> Array2<f32> {
-    arr.row_mut(row_id).assign(new_row);
-    arr
+/// `sigmoid` computes the sigmoid function for a scalar value.
+#[inline]
+pub fn sigmoid(x: f32) -> f32 {
+    1.0 / (1.0 + (-x).exp())
 }
 
-/// `concatenate_arr1` concatenates two 1d arrays to form another 1d array.
-pub fn concatenate_arr1(
-    arr1: ArrayBase<ViewRepr<&f32>, Dim<[usize; 1]>>,
-    arr2: ArrayBase<ViewRepr<&f32>, Dim<[usize; 1]>>,
-) -> Array1<f32> {
-    concatenate![Axis(0), arr1, arr2]
+/// Compute the dot product.
+///
+/// `xs` and `ys` must be the same length
+///
+/// (From ndarray 0.15.6)
+fn unrolled_dot(xs: &[f32], ys: &[f32]) -> f32 {
+    debug_assert_eq!(xs.len(), ys.len());
+    // eightfold unrolled so that floating point can be vectorized
+    // (even with strict floating point accuracy semantics)
+    let mut sum = 0.0;
+    let mut p = (0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0);
+    let xit = xs.chunks_exact(8);
+    let yit = ys.chunks_exact(8);
+    for (x, y) in xit.remainder().iter().zip(yit.remainder().iter()) {
+        sum += x * y;
+    }
+    for (xx, yy) in xit.zip(yit) {
+        // TODO: Use array_chunks once stable to avoid the unwrap.
+        // <https://github.com/rust-lang/rust/issues/74985>
+        #[allow(clippy::unwrap_used)]
+        let [x0, x1, x2, x3, x4, x5, x6, x7] = *<&[f32; 8]>::try_from(xx).unwrap();
+        #[allow(clippy::unwrap_used)]
+        let [y0, y1, y2, y3, y4, y5, y6, y7] = *<&[f32; 8]>::try_from(yy).unwrap();
+        p.0 += x0 * y0;
+        p.1 += x1 * y1;
+        p.2 += x2 * y2;
+        p.3 += x3 * y3;
+        p.4 += x4 * y4;
+        p.5 += x5 * y5;
+        p.6 += x6 * y6;
+        p.7 += x7 * y7;
+    }
+    sum += p.0 + p.4;
+    sum += p.1 + p.5;
+    sum += p.2 + p.6;
+    sum += p.3 + p.7;
+
+    sum
```

Comment:
consider
```suggestion
    sum + (p.0 + p.4) + (p.1 + p.5) + (p.2 + p.6) + (p.3 + p.7)
```

This way `sum` wouldn't have to be mutable.

Please also add a comment why the chosen order of operations is preferable over the unbracketed one.

Otherwise, if you implement Manish's suggestion of making `p: [f64; 8]`, this can be

```rust
    sum + p.iter().sum()
```